### ISSUE TITLE
Where is the import hook? / How does the import work?

### ISSUE THREAD

**Issue by Make42** [Nuitka/Nuitka#693]
After building my .pyd from my package, I encrypt it and save it as .epyd file on my disk (a binary file). I can read and decrypt the file with

```python
with open('src_nuitka/src.epyd', 'rb') as f:
     my_pyd_module = decrypt(f.read())
```

Now I would like to import the module using the `<class 'bytes'>` object `my_pyd_module` directly without writing to disk first. How can I do this? Since it is not a Python code string, I cannot use `exec`.

My idea is to write the file to an in-memory file system (fs), e.g. using [PyFilesystem](https://github.com/PyFilesystem/pyfilesystem) and then import it using a custom import hook. PyFilesystem provides an [import hook for their in-memory fs](https://pyfilesystem.readthedocs.io/en/latest/expose/importhook.html#) but it uses `exec`. One way or another Nuitka also has to import the pyd files and provide an import hook for that (I guess). So my idea is to combine the Nuitka import hook with the PyFilesystem import hook to read the pyd file from the in-memory fs.

My questions are thus:

1. Is there an option to load the .pyd file directy from `my_pyd_module`, so I do not have to write it anywhere? (Then I would not need PyFilesystem for this.) All examples of writing import hooks did this using `exec` or by instantiating the classes directly as in https://dev.to/dangerontheranger/dependency-injection-with-import-hooks-in-python-3-5hap.
2. How does the import of Nuitka work?
3. If Nuitka uses a custom import hook, where can I find the code for the import hook?

If you have any advice for the combination of the mentioned import hooks, please let me know.

**Comment by Make42** [comment]
I was able to solve the entire import hook issue, thanks to the posts

- https://blog.ffledgling.com/python-imports-i.html
- https://dev.to/dangerontheranger/dependency-injection-with-import-hooks-in-python-3-5hap
- https://xion.org.pl/2012/05/06/hacking-python-imports/
- https://www.python.org/dev/peps/pep-0302

However, as it turns out, I still need to be able to solve this issue, since import hooks still require to execute code. One way or another this problem needs to be solved for regular Nuitka as well, I would guess. But is the import hook for that maybe already implemented in standard python?

**Comment by kayhayen** [comment]
Nuitka implements a meta path based importer in C. It can load extension modules, bytecode, etc. as included. You will find in the source code as `nuitka/build/static_src/MetaPathBasedLoader.c` and you should be able to go from there.

**Comment by Make42** [comment]
Thank you! I know how how to write a meta path importer in Python, but I am not too firm in C. In Python I need to write the method `load_module` (PEP 302) among other things - or alternatively the method and `exec_module` in (PEP 451). I am guessing you are implementing `load_module` and your implementation is the function

    static PyObject *_path_unfreezer_load_module(PyObject *self, PyObject *args, PyObject *kwds)

right?

What I want to be able to do is to provide my Nuitka binary `pyd` as an object to the importer, substituting that the Nuitka importer is reading it from disk. So I guess I need to find the point where the Nuitka importer is reading from disk.

Within I followed `_path_unfreezer_load_module`, I am guessing that I have to follow to 

     IMPORT_EMBEDDED_MODULE(module_name, name)
     result = loadModule(module_name, entry);

where I do not go into `callIntoShlibModule(entry->name, filename);`, since I am not using a standalone, but [use-case 3](https://nuitka.net/doc/user-manual.html#use-case-3-package-compilation), but instead, the line

             PyCodeObject *code_object = (PyCodeObject *)PyMarshal_ReadObjectFromString(
            (char *)&constant_bin[entry->bytecode_start], entry->bytecode_size);

is the one I should be interested in. From https://docs.python.org/3/c-api/marshal.html I get that

>  PyObject* PyMarshal_ReadObjectFromString(const char *data, Py_ssize_t len)
    Return value: New reference.
>
>    Return a Python object from the data stream in a byte buffer containing len bytes pointed to by data.

so I am guessing that I need to give my data object (which I have already read) as the first argument to `PyMarshal_ReadObjectFromString`, which would be `(char *)&constant_bin[entry->bytecode_start]` in the Nuitka `loadModule`.

Ok, this is where I am lost again, since I am actually not able to write C. I understand that you are converting this "thing" to a pointer that points to a char and we get the value from `constant_bin[entry->bytecode_start]`, which, as I am guessing, is a pointer to the start of the binary object that I would like to provide.

I tried really hard to do my homework... am I on the right track? How do I go from here?
